src/hardware/sblaster.cpp

```diff
@@ -766,59 +766,68 @@ static void PlayDMATransfer(uint32_t bytes_requested)
-			MixTemp[samples++] = decode_ADPCM_2_sample((sb.dma.buf.b8[i] >> 2) & 0x3);
-			MixTemp[samples++] = decode_ADPCM_2_sample((sb.dma.buf.b8[i] >> 0) & 0x3);
-
-			frames = check_cast<uint16_t>(samples / channels);
-			sb.chan->AddSamples_m8(frames, maybe_silence(samples, MixTemp));
-			samples = 0;
+ 			const uint8_t decoded[] = {
+			        decode_ADPCM_2_sample((sb.dma.buf.b8[i] >> 6) & 0x3),
+			        decode_ADPCM_2_sample((sb.dma.buf.b8[i] >> 4) & 0x3),
+			        decode_ADPCM_2_sample((sb.dma.buf.b8[i] >> 2) & 0x3),
+			        decode_ADPCM_2_sample((sb.dma.buf.b8[i] >> 0) & 0x3)};
+
+			constexpr uint8_t num_decoded = ARRAY_LEN(decoded);
+			sb.chan->AddSamples_m8(num_decoded, maybe_silence(num_decoded, decoded));
+			samples += num_decoded;
 			++i;
 		}
+		frames = check_cast<uint16_t>(samples / channels);
+
 		break;
 	case DSP_DMA_3:
 		bytes_read = ReadDMA8(bytes_to_read);
 		if (bytes_read && sb.adpcm.haveref) {
 			sb.adpcm.haveref=false;
 			sb.adpcm.reference=sb.dma.buf.b8[0];
 			sb.adpcm.stepsize=MIN_ADAPTIVE_STEP_SIZE;
-			i++;
+			++i;
 		}
 		assert(samples == 0);
 		while (i < bytes_read) {
-			MixTemp[samples++] = decode_ADPCM_3_sample((sb.dma.buf.b8[i] >> 5) & 0x7);
-			MixTemp[samples++] = decode_ADPCM_3_sample((sb.dma.buf.b8[i] >> 2) & 0x7);
-			MixTemp[samples++] = decode_ADPCM_3_sample((sb.dma.buf.b8[i] & 0x3) << 1);
-
-			frames = check_cast<uint16_t>(samples / channels);
-			sb.chan->AddSamples_m8(frames, maybe_silence(samples, MixTemp));
-			samples = 0;
+			const uint8_t decoded[] = {
+			        decode_ADPCM_3_sample((sb.dma.buf.b8[i] >> 5) & 0x7),
+			        decode_ADPCM_3_sample((sb.dma.buf.b8[i] >> 2) & 0x7),
+			        decode_ADPCM_3_sample((sb.dma.buf.b8[i] & 0x3) << 1)};
+
+			constexpr uint8_t num_decoded = ARRAY_LEN(decoded);
+			sb.chan->AddSamples_m8(num_decoded, maybe_silence(num_decoded, decoded));
+			samples += num_decoded;
 			++i;
 		}
+		frames = check_cast<uint16_t>(samples / channels);
+
 		break;
 	case DSP_DMA_4:
 		bytes_read = ReadDMA8(bytes_to_read);
 		if (bytes_read && sb.adpcm.haveref) {
 			sb.adpcm.haveref=false;
 			sb.adpcm.reference=sb.dma.buf.b8[0];
 			sb.adpcm.stepsize=MIN_ADAPTIVE_STEP_SIZE;
-			i++;
+			++i;
 		}
 		assert(samples == 0);
 		while (i < bytes_read) {
-			MixTemp[samples++] = decode_ADPCM_4_sample(sb.dma.buf.b8[i] >> 4);
-			MixTemp[samples++] = decode_ADPCM_4_sample(sb.dma.buf.b8[i] & 0xf);
+			const uint8_t decoded[] = {
+			        decode_ADPCM_4_sample(sb.dma.buf.b8[i] >> 4),
+			        decode_ADPCM_4_sample(sb.dma.buf.b8[i] & 0xf)};
 
-			frames = check_cast<uint16_t>(samples / channels);
-			sb.chan->AddSamples_m8(frames, maybe_silence(samples, MixTemp));
-			samples = 0;
+			constexpr uint8_t num_decoded = ARRAY_LEN(decoded);
+			sb.chan->AddSamples_m8(num_decoded, maybe_silence(num_decoded, decoded));
+			samples += num_decoded;
 			++i;
 		}
+		frames = check_cast<uint16_t>(samples / channels);
+
```

Comment:
So what fixed the bug? I'm sure you tested it and it works, I'm just not seeing it from these changes, so just curious :)

Reply:
(oh; I forgot to flag it!) 
In the three while loops, previously the `samples = ` were reset to zero instead of holding the total for the given DMA transfer, which is where the accounting fell apart. `frames`, on the other hand, were correctly tallied based on the intermediate `samples` count, and so that's why frames was larger than samples.

Reply:
Makes sense, i suspected something like that.